File: src/chain-observers/filters/account-created-filter.ts
import type { WorkerBee } from "../../bot";
import { OperationClassifier } from "../classifiers";
import type { TRegisterEvaluationContext } from "../classifiers/collector-classifier-base";
import type { TFilterEvaluationContext } from "../factories/data-evaluation-context";
import { FilterBase } from "./filter-base";

export class AccountCreatedFilter extends FilterBase {
  public constructor(
    worker: WorkerBee
  ) {
    super(worker);
  }

  public usedContexts(): Array<TRegisterEvaluationContext> {
    return [
      OperationClassifier
    ];
  }

  public async match(data: TFilterEvaluationContext): Promise<boolean> {
    const { operationsPerType } = await data.get(OperationClassifier);

    for(const {} of (operationsPerType.account_create ?? []))
      return true;

    for(const {} of (operationsPerType.account_create_with_delegation ?? []))
      return true;

    for(const {} of (operationsPerType.create_claimed_account ?? []))
      return true;

    return false;
  }
}

File: python/workerbee/chain_observers/filters/account_created_filter.py
"""AccountCreatedFilter — matches account creation operations.

Mirrors src/chain-observers/filters/account-created-filter.ts.

Uses OperationClassifier. Matches account_create / account_create_with_delegation /
create_claimed_account operations.
"""

from __future__ import annotations

from typing import TYPE_CHECKING

from .filter_base import FilterBase

if TYPE_CHECKING:
    from ..classifiers.collector_classifier_base import TRegisterEvaluationContext
    from ..factories.data_evaluation_context import DataEvaluationContext


class AccountCreatedFilter(FilterBase):
    def used_contexts(self) -> list[TRegisterEvaluationContext]:
        from ..classifiers.operation_classifier import OperationClassifier

        return [OperationClassifier]

    async def match(self, data: DataEvaluationContext) -> bool:
        from ..classifiers.operation_classifier import OperationClassifier

        result = await data.get(OperationClassifier)
        operations_per_type = result["operations_per_type"]

        return bool(
            operations_per_type.get("account_create_operation")
            or operations_per_type.get("account_create_with_delegation_operation")
            or operations_per_type.get("create_claimed_account_operation"),
        )
